### PR TITLE
Rewrite of DIP1006 based on community feedback

### PR DESCRIPTION
**Pull request by leandro-lucarella-sociomantic**
This is only made as a PR for preliminary review.

### CODE REVIEW

File: DIPs/DIP1006.md [hunk 1]
```diff
@@ -1,111 +1,306 @@
...
+checks (`invariant`, `in`, `out`, `assert`, bounds checking and unexpected cases
+in `switch`) and to unify the existing ways to customize run-time checks
+(`-release` and `-boundscheck`).
+
+### Reference
+
+* Limited D1 implementation, used by Sociomantic:
+  https://github.com/dlang/dmd/pull/6347
+
+* Similar proposal in the form of a PR by Martin Novak:
```
Comment (gautam-kotian-sociomantic) [inline DIPs/DIP1006.md]: `s/Novak/Nowak`

File: DIPs/DIP1006.md [hunk 2]
```diff
@@ -1,111 +1,306 @@
...
+
+Profiling some real-time applications within Sociomantic showed that
+`_d_invariant` was the most expensive call in the application. Affected
+applications were using few or no invariants, but since Sociomantic code is
+mostly written in an OOP style, simply disabling invariants (with no other
+optimization) led to 20% more throughput.
+
+### Summary of the current issues
+
+* Not all user code needs the same level of safety in expense of performance, so
```
Comment (gautam-kotian-sociomantic) [inline DIPs/DIP1006.md]: `s/in expense of/at the cost of`

File: DIPs/DIP1006.md [hunk 3]
```diff
@@ -1,111 +1,306 @@
...
 
-#### Relationship with `-release` and `-debug`
+Because of this, the presence of `invariant` code in the object file/binary
+is subject to  both `invariant` and `assert` checks.
 
-In DMD, `-debug` just mean enabling `debug` blocks, so there is no overlap with `-contracts`'s proposed functionality.
-On the other hand, `-release` disables all aforementioned contracts, and, in addition, disables "switch errors"
-(compiler-generated `default` case in a switch statement are turned into HLT instruction instead of throwing a `SwitchError`).
-As a result, using `-contracts` with `-release` is equivalent to just using `-release`.
+If `-check=-invariant,+assert` is used, the `invariant` function will still be
```
Comment (gautam-kotian-sociomantic) [inline DIPs/DIP1006.md]: Should be `-checks=...`

File: DIPs/DIP1006.md [hunk 4]
```diff
@@ -1,111 +1,306 @@
...
+
+#### `-debug`
+
+The `-checks` option doesn't interfere at all with the `-debug` option, which
+only enables `debug` blocks. `debug` blocks that are compiled in when using
+`-debug` will be subject to the same checks specified by `-checks` as any other
+parts of the user code.
+
+In particular, if they contain `assert()` calls and `-assert` is specified, those
+`assert()` calls wont run, but all other code in the `debug` block will. It is up
```
Comment (gautam-kotian-sociomantic) [inline DIPs/DIP1006.md]: `s/wont/won't`

File: DIPs/DIP1006.md [hunk 5]
```diff
@@ -1,111 +1,306 @@
...
+
+The `-checks` option doesn't interfere at all with the `-unittest` option, which
+only enables `unittest` blocks (compiles them and, depending on the `druntime`
+unit test runner, runs them as part of the program's initialization).
+
+`unittest` blocks that are compiled in when using `-unittest` will be subject to
+the same checks specified by `-checks` as any other parts of the user code.
+
+In particular, if they contain `assert()` calls and `-assert` is specified,
+those `assert()` calls wont run, but all other code in the `unittest` block
```
Comment (gautam-kotian-sociomantic) [inline DIPs/DIP1006.md]: `s/wont/won't`

**Review by gautam-kotian-sociomantic — COMMENTED** [review]
LGTM apart from the minor comments.

I was reviewing one commit at a time, which meant many of my comments were automatically fixed in a later commit that included Mike Parker's changes :(

May be worth squashing before calling for more reviews so that no one else makes the same mistake.

**Comment by leandro-lucarella-sociomantic** [comment]
Fixed, thanks! I will close for now because another round of big changes are coming as a result from the official review with Andrei.

**Comment by leandro-lucarella-sociomantic** [comment]
> I was reviewing one commit at a time, which meant many of my comments were automatically fixed in a later commit that included Mike Parker's changes :(

Sorry about that, I think is better to keep the history in this case for the sake of transparency. But I will make sure to clarify this is the case when I open a new PR for review, so people can first give the whole list of commits a look and then do the review on the consolidated diff instead of in a commit-by-commit basis if that suits this case better.

**Pull request closed without merging**